tests/snap_test.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil || stderr != "" {
+		log.Printf("stderr: %s", stderr)
+		log.Printf("err: %s", err)
+		return "", err
+	}
+
+	utils.Exec(nil, "rm -rf squashfs-root")
+	return newPath, nil
+}
+
+func setup() (teardown func(), err error) {
+	var newPath string
+
+	log.Println("[CLEAN]")
+	utils.SnapRemove(nil, snapMatterPiGPIO)
+	utils.SnapRemove(nil, chipToolSnap)
+
+	log.Println("[SETUP]")
+
+	teardown = func() {
+		log.Println("[TEARDOWN]")
+		utils.SnapDumpLogs(nil, start, snapMatterPiGPIO)
+
+		utils.Exec(nil, "rm "+newPath)
+
+		log.Println("Removing installed snap:", !utils.SkipTeardownRemoval)
+		if !utils.SkipTeardownRemoval {
+			utils.SnapRemove(nil, snapMatterPiGPIO)
+			utils.SnapRemove(nil, chipToolSnap)
+		}
+	}
+
+	// authorize gpio mock
+	if utils.LocalServiceSnap() {
+		newPath, err = authorizeGpioMock(utils.LocalServiceSnapPath)
+	}
+	if err != nil {
+		teardown()
+		return
+	}
+
+	// install matter-pi-gpio-commander
+	if utils.LocalServiceSnap() {
+		err = utils.SnapInstallFromFile(nil, newPath)
+	} else {
+		err = utils.SnapInstallFromStore(nil, snapMatterPiGPIO, utils.ServiceChannel)
+	}
+	if err != nil {
+		teardown()
+		return
+	}
+
+	if err = setupGPIO(); err != nil {
+		teardown()
+		return
+	}
+
+	// connect interfaces:
+	utils.SnapConnect(nil, snapMatterPiGPIO+":avahi-control", "")
+	utils.SnapConnect(nil, snapMatterPiGPIO+":bluez", "")
+	utils.SnapConnect(nil, snapMatterPiGPIO+":network", "")
+	utils.SnapConnect(nil, snapMatterPiGPIO+":network-bind", "")
+	utils.SnapConnect(nil, snapMatterPiGPIO+":custom-gpio", snapMatterPiGPIO+":custom-gpio-dev")
+
+	return
+}
+
+func getMockGPIO() (string, error) {
+	gpioChipNumber, stderr, err := utils.Exec(nil, "ls /dev/gpiochip* | sort -n | tail -n 1 | cut -d'p' -f3")
+	if err != nil || stderr != "" {
+		log.Printf("Failed to get mock gpio chip number, Error %s: %s", stderr, err)
+		return "", err
+	}
+	return gpioChipNumber, nil
+}
+
+func setupGPIO() error {
+
+	gpioChip := os.Getenv(specificGpioChip)
+	var err error
+	if gpioChip == "" {
+		gpioChip, err = getMockGPIO()
+		if err != nil {
+			return fmt.Errorf("failed to get mock gpio chip number: %s", err)
+		}
+
+		log.Printf("[TEST] No specific gpiochip defined, using mockup gpio: /dev/gpiochip%s", gpioChip)
+	}
+
+	gpioLine := os.Getenv(specifGpioLine)
+	if gpioLine == "" {
+		gpioLine = "4"
+		log.Printf("[TEST] No specific gpio line defined, using default: %s", gpioLine)
+	}
+
+	utils.SnapSet(nil, snapMatterPiGPIO, "gpiochip", gpioChip)
+	utils.SnapSet(nil, snapMatterPiGPIO, "gpio", gpioLine)
+
+	return nil
+}
+
+func TestBlinkOperation(t *testing.T) {
+	// test blink operation
+	ctx, cancel := context.WithDeadline(context.Background(), <-time.After(20*time.Second))
+	defer cancel()
+
+	stdout, _, err := utils.ExecContextVerbose(nil, ctx, snapMatterPiGPIO+".test-blink")
+	t.Logf("err: %s", err)
+	t.Logf("stdout: %s", stdout)
+	assert.Error(t, err)
+	assert.True(t, errors.Is(err, context.DeadlineExceeded))
```

Comment:
This is testing the `utils.ExecContextVerbose` function rather than the test-blink application. The tests should check and make sure the application logs contain the expected output and none of the expected gpio or config errors.

Reply:
Fixed! Thank you! Added assertions to check the log output!